spacy/cli/project/parallel.py

```diff
@@ -0,0 +1,442 @@
+    # group is not allowed to contain the same command more than once
+    log_file_name = os.sep.join(
+        (current_dir, PARALLEL_LOGGING_DIR_NAME, cmd_info.cmd_name + ".log")
+    )
+    file_not_found = False
+    # buffering=0: make sure output is not lost if a subprocess is terminated
+    with open(log_file_name, "wb", buffering=0) as logfile:
+        for os_cmd_index, os_cmd in enumerate(cmd_info.cmd["script"]):
+            command = split_command(os_cmd)
+            if len(command) and command[0] in ("python", "python3"):
+                # -u: prevent buffering within Python
+                command = [sys.executable, "-u", *command[1:]]
+            elif len(command) and command[0] in ("pip", "pip3"):
+                command = [sys.executable, "-m", "pip", *command[1:]]
+            logfile.write(
+                bytes(
+                    f"Running command: {join_command(command)}" + os.linesep,
+                    encoding="utf8",
+                )
+            )
+            if not dry:
+                try:
+                    sp = Popen(
+                        command,
+                        stdout=logfile,
+                        stderr=STDOUT,
+                        env=os.environ.copy(),
+                        encoding="utf8",
+                    )
+                except FileNotFoundError:
+                    # Indicates the *command* wasn't found, it's an error before the command
+                    # is run.
+                    logfile.write(
+                        bytes(
+                            Errors.E970.format(
+                                str_command=" ".join(command),
+                                tool=command[0],
+                            ),
+                            encoding="UTF-8",
+                        )
+                    )
+                    file_not_found = True
+                    break
+                parallel_group_status_queue.put(
+                    {
+                        "type": "started",
+                        "cmd_index": cmd_info.cmd_index,
+                        "os_cmd_index": os_cmd_index,
+                        "pid": sp.pid,
+                    }
+                )
+                while True:
+                    try:
+                        sp.wait(timeout=PARALLEL_GROUP_STATUS_INTERVAL)
+                    except TimeoutExpired:
+                        pass
+                    if sp.returncode == None:
+                        parallel_group_status_queue.put(
+                            {
+                                "type": "keepalive",
+                                "cmd_index": cmd_info.cmd_index,
+                            }
+                        )
+                    else:
+                        break
+
+                if sp.returncode != 0:
+                    if os.name == "nt":
+                        status = "Failed/terminated"
```

Comment:
If we have overlapping status values, we could also consider bit masking. If this is the only case though (I think it is?), then it doesn't make sense/would be overengineering.

Reply:
This is the only case, so I think bit masking would be too much.